crypto/hpke.c

```diff
@@ -0,0 +1,2514 @@
+        memcpy(kem_context, key1enc, key1enclen);
+        memcpy(kem_context + key1enclen, key2enc, key2enclen);
+    } else {
+        memcpy(kem_context, key2enc, key2enclen);
+        memcpy(kem_context + key2enclen, key1enc, key1enclen);
+    }
+    if (apublen != 0) {
+        /* Append the public auth key (mypub) to kem_context */
+        if ((kem_contextlen + apublen) >= HPKE_MAXSIZE) {
+            HPKE_err;
+        }
+        memcpy(kem_context + kem_contextlen, apub, apublen);
+        kem_contextlen += apublen;
+    }
+
+    if (akey != NULL) {
+        size_t zzlen2 = 0;
+
+        /* step 2 run to get 2nd half of zz */
+        if (encrypting) {
+            pctx = EVP_PKEY_CTX_new_from_pkey(libctx, akey, NULL);
+        } else {
+            pctx = EVP_PKEY_CTX_new_from_pkey(libctx, key1, NULL);
+        }
+        if (pctx == NULL) {
+            HPKE_err;
+        }
+        if (EVP_PKEY_derive_init(pctx) <= 0) {
+            HPKE_err;
+        }
+        if (encrypting) {
+            if (EVP_PKEY_derive_set_peer(pctx, key2) <= 0) {
+                HPKE_err;
+            }
+        } else {
+            if (EVP_PKEY_derive_set_peer(pctx, akey) <= 0) {
+                HPKE_err;
+            }
+        }
+        if (EVP_PKEY_derive(pctx, NULL, &zzlen2) <= 0) {
+            HPKE_err;
+        }
+        if (zzlen2 >= HPKE_MAXSIZE) {
+            HPKE_err;
+        }
+        if (EVP_PKEY_derive(pctx, zz + zzlen, &zzlen2) <= 0) {
+            HPKE_err;
+        }
+        zzlen += zzlen2;
+        EVP_PKEY_CTX_free(pctx);
+        pctx = NULL;
+    }
+    erv = hpke_extract_and_expand(libctx, suite, HPKE_5869_MODE_KEM,
+                                  zz, zzlen, kem_context, kem_contextlen,
+                                  lss, &lsslen);
+    if (erv != 1) { goto err; }
+    *ss = OPENSSL_malloc(lsslen);
+    if (*ss == NULL) {
+        HPKE_err;
+    }
+    memcpy(*ss, lss, lsslen);
+    *sslen = lsslen;
+
+err:
+    if (pctx != NULL)
+        EVP_PKEY_CTX_free(pctx);
+    return erv;
+}
+
+/*
+ * @brief check mode is in-range and supported
+ * @param mode is the caller's chosen mode
+ * @return 1 for good (OpenSSL style), not 1 for error
+ */
+static int hpke_mode_check(unsigned int mode)
+{
+    switch (mode) {
+    case HPKE_MODE_BASE:
+    case HPKE_MODE_PSK:
+    case HPKE_MODE_AUTH:
+    case HPKE_MODE_PSKAUTH:
+        break;
+    default:
+        return (__LINE__);
+    }
+    return (1);
+}
+
+/*
+ * @brief check psk params are as per spec
+ * @param mode is the mode in use
+ * @param pskid PSK identifier
+ * @param psklen length of PSK
+ * @param psk the psk itself
+ * @return 1 for good (OpenSSL style), not 1 for error
+ *
+ * If a PSK mode is used both pskid and psk must be
+ * non-default. Otherwise we ignore the PSK params.
+ */
+static int hpke_psk_check(unsigned int mode,
+                          char *pskid,
+                          size_t psklen,
+                          unsigned char *psk)
```

Comment:
could be `const unsigned char*` argument

Reply:
sure, did that (again:-)